ls1intum/Artemis
# `Assessment`: Fix issues in the calculation of participant scores

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context



**Background:**
_Participant scores_ save the last (rated) _result_ for each combination of student/team and exercise.
Otherwise, we would have to go through the whole relation chain `Exercise -> Participation -> Submission -> Result` to find the last result for each student just to calculate the average course score for an exercise.

**Problem:**
Currently, we rely on callback methods from JPA entity lifecycle events (`@PostPersist`, `@PreRemove`, …). Then, we perform some database queries to fetch the participation and find the applicable last _result_.
However, Hibernate advises that [JPA callback methods](https://docs.jboss.org/hibernate/orm/5.0/userguide/en-US/html/ch12.html#d5e1611) must “not invoke EntityManager or Query methods!”. A recent issue #5633 demonstrates that this solution uses undefined behavior and may break easily. Also, every time a result is changed, the _participant score_ calculation runs and therefore slows down all requests modifying a result entity. Lastly, if the instance shuts down or an issue occurs during the calculation, the _participant score_ table does not repair itself and the value becomes invalid.

### Description


We move the calculation of _participant scores_ to a scheduler. The JPA callback methods send a message to the main instance to schedule the update of a _participant score_ for a modified _result_.
This ensures that the _participant scores_ are updated promptly within seconds.
We introduce a `last_modified_date` column for both _results_ and _participant scores_.
Using this date, a regular cron job schedules the processing of modified results. This also means that if the main instance goes down, the _participant scores_ can heal themselves the next time it boots up.

All in all, regardless of what happens, the _participant scores_ should eventually be consistent.

### Steps for Testing

Prerequisites:
- 1 Instructor, 1 Student
- 1 Course
- 1 Exercise of each type
- 1 Exam with an exercise of each type

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to the Course and open the _"Participant Scores"_ in a new tab (**ensure it stays up-to-date for all tests below**)
4. Participate in each exercise as student
5. Assess the exercise as instructor 
6. Also try: Delete/Reset or edit the assessment, submission, participation or exercise
7. Participate in the exam as a student
8. Assess the exam as instructor 

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Instructions | Branches |
|------------|-------:|-----:|
| ResultListener.java | 94% | 75% |
| ParticipantScoreSchedulerService.java | 92% | 77% |
